src/main/java/com/hivemq/client/internal/mqtt/handler/publish/outgoing/MqttOutgoingQosHandler.java

```diff
@@ -121,32 +123,36 @@ public void onSessionStartOrResume(final @NotNull MqttClientConnectionConfig con
                     .subscribe(this);
             assert subscription != null;
             packetIdentifiers = new Ranges(1, newSendMaximum);
-            qos1Or2Map = IntMap.range(1, newSendMaximum);
+            pending = IntMap.range(1, newSendMaximum);
             subscription.request(newSendMaximum);
         } else {
             assert packetIdentifiers != null;
-            assert qos1Or2Map != null;
+            assert pending != null;
             assert subscription != null;
             resize();
             final int newRequests = newSendMaximum - oldSendMaximum - shrinkRequests;
             if (newRequests > 0) {
-                subscription.request(newRequests);
                 shrinkRequests = 0;
+                subscription.request(newRequests);
             } else {
                 shrinkRequests = -newRequests;
             }
-//            resend(); // TODO
         }
         topicAliasMapping = connectionConfig.getSendTopicAliasMapping();
+
+        if ((firstPending != null) || queuedCounter.get() > 0) {
+            resendPending = firstPending;
+            eventLoop.execute(this);
+        }
     }
 
     private void resize() {
         assert packetIdentifiers != null;
-        assert qos1Or2Map != null;
+        assert pending != null;
 
         shrinkIds = packetIdentifiers.resize(sendMaximum);
         if (shrinkIds == 0) {
-            qos1Or2Map = IntMap.resize(qos1Or2Map, sendMaximum);
+            pending = IntMap.resize(pending, sendMaximum);
```

Comment:
Are onError and anComplete the right log statements? "MqttPublishFlowables..."?

Reply:
Do you mean onComplete/onError must not be called? This is true

Reply:
Why MqttPublishFlowables tho and not MqttOutgoingQosHandler?

Reply:
The MqttOutgoingQosHandler ist just the subscriber to the MqttPublishFlowables. These must not error/complete so the callback methods onError/onComplete must not be called

Reply:
Ah ok.